Reproducible Failure in org.elasticsearch.search.aggregations.bucket.TermsDocCountErrorIT.testIncrementalReduction

With current master:

```
./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.search.aggregations.bucket.TermsDocCountErrorIT.testIncrementalReduction" -Dtests.seed=370A19D46176CFA7 -Dtests.locale=zh-SG -Dtests.timezone=Atlantic/Madeira -Druntime.java=16
```

fails 100% of the time with 

```
  2> REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.search.aggregations.bucket.TermsDocCountErrorIT.testIncrementalReduction" -Dtests.seed=370A19D46176CFA7 -Dtests.locale=zh-SG -Dtests.timezone=Atlantic/Madeira -Druntime.java=16
  2> java.lang.AssertionError: 
    Expected: <0L>
         but: was <1L>
        at __randomizedtesting.SeedInfo.seed([370A19D46176CFA7:1A95D442705B6E10]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.elasticsearch.search.aggregations.bucket.TermsDocCountErrorIT.testIncrementalReduction(TermsDocCountErrorIT.java:1048)

```

for me. Also fails for a bunch of other seeds as well.

--- a/benchmarks/src/main/java/org/elasticsearch/benchmark/search/aggregations/TermsReduceBenchmark.java
+++ b/benchmarks/src/main/java/org/elasticsearch/benchmark/search/aggregations/TermsReduceBenchmark.java
@@ -154,7 +154,7 @@ public class TermsReduceBenchmark {
                 true,
                 0,
                 buckets,
-                0L
+                null
             );
         }
 

--- a/benchmarks/src/main/java/org/elasticsearch/benchmark/search/aggregations/bucket/terms/StringTermsSerializationBenchmark.java
+++ b/benchmarks/src/main/java/org/elasticsearch/benchmark/search/aggregations/bucket/terms/StringTermsSerializationBenchmark.java
@@ -70,7 +70,7 @@ public class StringTermsSerializationBenchmark {
             false,
             100000,
             resultBuckets,
-            0L
+            null
         );
     }
 

--- a/server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/bucket/TermsDocCountErrorIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/bucket/TermsDocCountErrorIT.java
@@ -1034,7 +1034,6 @@ public class TermsDocCountErrorIT extends ESIntegTestCase {
      * Tests the upper bounds are correct when performing incremental reductions
      * See https://github.com/elastic/elasticsearch/issues/40005 for more details
      */
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/75667")
     public void testIncrementalReduction() {
         SearchResponse response = client().prepareSearch("idx_fixed_docs_3", "idx_fixed_docs_4", "idx_fixed_docs_5")
             .addAggregation(terms("terms")

--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/AbstractInternalTerms.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/AbstractInternalTerms.java
@@ -340,7 +340,7 @@ public abstract class AbstractInternalTerms<
 
     protected static XContentBuilder doXContentCommon(XContentBuilder builder,
                                                       Params params,
-                                                      long docCountError,
+                                                      Long docCountError,
                                                       long otherDocCount,
                                                       List<? extends AbstractTermsBucket> buckets) throws IOException {
         builder.field(DOC_COUNT_ERROR_UPPER_BOUND_FIELD_NAME.getPreferredName(), docCountError);

--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/AbstractInternalTerms.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/AbstractInternalTerms.java
@@ -133,7 +133,7 @@ public abstract class AbstractInternalTerms<
         if (size == 0 || size < terms.getShardSize() || isKeyOrder(terms.getOrder())) {
             return 0;
         } else if (InternalOrder.isCountDesc(terms.getOrder())) {
-            if (terms.getDocCountError() != null && terms.getDocCountError() > 0) {
+            if (terms.getDocCountError() != null) {
                 // If there is an existing docCountError for this agg then
                 // use this as the error for this aggregation
                 return terms.getDocCountError();

--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java
@@ -765,7 +765,7 @@ public class GlobalOrdinalsStringTermsAggregator extends AbstractStringTermsAggr
             }
             return new StringTerms(name, reduceOrder, order, bucketCountThresholds.getRequiredSize(),
                 bucketCountThresholds.getMinDocCount(), metadata(), format, bucketCountThresholds.getShardSize(), showTermDocCountError,
-                otherDocCount, Arrays.asList(topBuckets), 0L);
+                otherDocCount, Arrays.asList(topBuckets), null);
         }
 
         @Override

--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedTerms.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedTerms.java
@@ -53,12 +53,13 @@ public abstract class InternalMappedTerms<A extends InternalTerms<A, B>, B exten
     protected InternalMappedTerms(StreamInput in, Bucket.Reader<B> bucketReader) throws IOException {
         super(in);
         if (in.getVersion().onOrAfter(Version.V_8_0_0)) { // todo fix after backport
-            docCountError = in.readOptionalLong();
-        } else {
-            docCountError = in.readZLong();
-            if (docCountError == 0) {
+            if (in.readBoolean()) {
+                docCountError = in.readZLong();
+            } else {
                 docCountError = null;
             }
+        } else {
+            docCountError = in.readZLong();
         }
         format = in.readNamedWriteable(DocValueFormat.class);
         shardSize = readSize(in);
@@ -70,7 +71,12 @@ public abstract class InternalMappedTerms<A extends InternalTerms<A, B>, B exten
     @Override
     protected final void writeTermTypeInfoTo(StreamOutput out) throws IOException {
         if (out.getVersion().onOrAfter(Version.V_8_0_0)) {  // todo fix after backport
-            out.writeOptionalLong(docCountError);
+            if (docCountError != null) {
+                out.writeBoolean(true);
+                out.writeZLong(docCountError);
+            } else {
+                out.writeBoolean(false);
+            }
         } else {
             out.writeZLong(docCountError == null ? 0 : docCountError);
         }

--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/MapStringTermsAggregator.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/MapStringTermsAggregator.java
@@ -446,7 +446,7 @@ public class MapStringTermsAggregator extends AbstractStringTermsAggregator {
             }
             return new StringTerms(name, reduceOrder, order, bucketCountThresholds.getRequiredSize(),
                 bucketCountThresholds.getMinDocCount(), metadata(), format, bucketCountThresholds.getShardSize(), showTermDocCountError,
-                otherDocCount, Arrays.asList(topBuckets), 0L);
+                otherDocCount, Arrays.asList(topBuckets), null);
         }
 
         @Override

--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/NumericTermsAggregator.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/NumericTermsAggregator.java
@@ -372,7 +372,7 @@ public class NumericTermsAggregator extends TermsAggregator {
                 showTermDocCountError,
                 otherDocCount,
                 List.of(topBuckets),
-                0L
+                null
             );
         }
 
@@ -454,7 +454,7 @@ public class NumericTermsAggregator extends TermsAggregator {
                 showTermDocCountError,
                 otherDocCount,
                 List.of(topBuckets),
-                0L
+                null
             );
         }
 

--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTermsAggregatorFromFilters.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTermsAggregatorFromFilters.java
@@ -227,7 +227,7 @@ public class StringTermsAggregatorFromFilters extends AdaptingAggregator {
             showTermDocCountError,
             otherDocsCount,
             buckets,
-            0L
+            null
         );
     }
 

--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/UnmappedTerms.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/UnmappedTerms.java
@@ -97,7 +97,7 @@ public class UnmappedTerms extends InternalTerms<UnmappedTerms, UnmappedTerms.Bu
 
     @Override
     public final XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
-        return doXContentCommon(builder, params, 0, 0, Collections.emptyList());
+        return doXContentCommon(builder, params, 0L, 0, Collections.emptyList());
     }
 
     @Override

